src/compiler/scanner.ts

```diff
@@ -2469,55 +2468,92 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
                 else if (ch === CharacterCodes.closeBracket) {
                     inCharacterClass = false;
                 }
-                p++;
+                pos++;
             }
-            const isUnterminated = !!(tokenFlags & TokenFlags.Unterminated);
-            const endOfBody = p - (isUnterminated ? 0 : 1);
-            let regExpFlags = RegularExpressionFlags.None;
-            while (p < end) {
-                const ch = charCodeUnchecked(p);
-                if (!isIdentifierPart(ch, languageVersion)) {
-                    break;
-                }
-                if (reportErrors) {
-                    const flag = characterToRegularExpressionFlag(String.fromCharCode(ch));
-                    if (flag === undefined) {
-                        error(Diagnostics.Unknown_regular_expression_flag, p, 1);
+            const endOfRegExpBody = pos;
+            if (tokenFlags & TokenFlags.Unterminated) {

```

Comment:
Oh sorry I should’ve explained this sooner and clearer. This section is the whole objective of the PR.
Since unterminated regexes are simply invalid that they don’t even form a complete RegularExpressionLiteral, we can do everything at our discretion as long as a syntax error is produced. But we should do all our best to parse the remaining parts with fewer subsequent syntax errors.

Instead of consuming all until the end of the line starting from the slash character, if we encounter an unbalanced bracket, which is likely a bracket balancing something before the regex instead of inside the regex, we stop scanning the regex here such as to increase the chance that a bracket forms a pair of syntax characters with the corresponding bracket before the regex.

Reply:
Thank you for the explanation. My primary concern is the allocation of the `bracketStack` array, as it introduces an array allocation per scanned RegExp, even when that RegExp contains no brackets, parens, or braces. 

It might be possible to write this in a way such that an allocation isn't necessary. Parentheses can nest, and square brackets can nest, but square brackets can't have nested parentheses, and curly braces cannot have nested parentheses, braces, or brackets.

Instead, you can count the number of left parentheses until you encounter either a left bracket, left brace, or right parenthesis. If you encounter a left bracket, you then count the number of left brackets until you encounter an equal number of right brackets. If you encounter a left brace, you scan for the next right brace. If you encounter a right parenthesis, you decrement the left parenthesis count. Instead of an array, you only really need two counters and a `boolean`.

Reply:
> curly braces cannot have nested parentheses, braces, or brackets

It is not the case in Annex B. That’s why I wrote like this.

Reply:
> > curly braces cannot have nested parentheses, braces, or brackets
> 
> It is not the case in Annex B. That’s why I wrote like this.

Can you point to where? Per the Annex B grammar, the only places `{` is balanced are [Quantifier](https://tc39.es/ecma262/#prod-Quantifier) and [InvalidBracedQuantifier](https://tc39.es/ecma262/#prod-annexB-InvalidBracedQuantifier), both of which only allow [DecimalDigits](https://tc39.es/ecma262/#prod-DecimalDigits) and comma (`,`). Otherwise, it matches [ExtendedPatternCharacter](https://tc39.es/ecma262/#prod-annexB-ExtendedPatternCharacter), which isn't balanced.

Reply:
Well, what I mean is regexes like `/{{\s*(\w+)\s*}}` do exist in the wild.

Reply:
Maybe so, but those braces don't need to be matched. If you encounter an open brace that doesn't match a quantifier, then you don't need to balance it.